src/MongoDB.Driver.Core/Core/Clusters/IClusterExtensions.cs

```diff
@@ -0,0 +1,89 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Bindings;
+using MongoDB.Driver.Core.Clusters.ServerSelectors;
+using MongoDB.Driver.Core.Servers;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    /// <summary>
+    /// This class contains extension methods for ICluster used for server selection with sharded transactions.
+    /// </summary>
+    internal static class IClusterExtensions
+    {
+        public static IServer SelectServerAndPinIfNeeded(
+            this ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            if (TryGetOrClearPinnedShardRouter(cluster, session, out var pinnedServer))
+            {
+                return pinnedServer;
+            }
+
+            // Server selection also updates the cluster type, allowing us to to determine if the shard router
+            // should be pinned.
+            var server = cluster.SelectServer(selector, cancellationToken);
+            PinShardRouterIfNeeded(cluster, session, server);
+            return server;
+        }
+        
+        public static async Task<IServer> SelectServerAndPinIfNeededAsync(
+            this ICluster cluster,
+            ICoreSessionHandle session,
+            IServerSelector selector,
+            CancellationToken cancellationToken)
+        {
+            if (TryGetOrClearPinnedShardRouter(cluster, session, out var pinnedServer))
+            {
+                return pinnedServer;
+            }
+
+            // Server selection also updates the cluster type, allowing us to to determine if the shard router
+            // should be pinned.
+            var server = await cluster.SelectServerAsync(selector, cancellationToken).ConfigureAwait(false);
+            PinShardRouterIfNeeded(cluster, session, server);
+
+            return server;
+        }
+        
+        private static void PinShardRouterIfNeeded(ICluster cluster, ICoreSessionHandle session, IServer server)
+        {
+            if (cluster.Description.Type == ClusterType.Sharded && session.IsInTransaction)
+            {
+                session.PinnedShardRouter = server;
+            }
+        }
+
+        private static bool TryGetOrClearPinnedShardRouter(ICluster cluster, ICoreSessionHandle session, out IServer pinnedServer)
+        {
+            var clusterType = cluster.Description.Type; 
```

Comment:
I still think this is a problem.  Consider a simple use case like:

    // 1. create client
    // 2. start session
    // 3. start transaction
    // 4. do operation in transaction

It's very likely that the cluster type is still unknown by the time the application reaches step 4.  

I think we have to make this a bit more complicated to handle that.  Essentially, we have to block until the cluster type is known.  One way to do that is by doing a SelectServer with a ServerSelector that matches any connected server.

Reply:
I don't think this is the problem you think it is. In your scenario this method will return false when the cluster type is unknown and the caller will immediately call SelectServer.

Reply:
I think you're right.  But in that case can't the conditional just be:

     if (session.PinnedShardRouter != null)

?  I also find it confusing that we call ICluster.SelectServer in all cases and then sometimes just ignore the result. It makes the intent confusing.

Reply:
Scratch that last part.  I was thinking of the Java implementation.

Reply:
And for that matter why not just return the pinned server, and just check for null in the caller, rather than having a boolean + out param?